### ISSUE TITLE
GLSLang does not error on invalid push constant vector indexing.

### ISSUE THREAD

**Issue by reduz** [KhronosGroup/glslang#1809]
I found that, for this piece of push constant:

```
layout(push_constant, binding = 0, std430) uniform DrawData {
    vec2 world_x;
    vec2 world_y;
    vec2 world_ofs;
    uint flags;
    uint specular_shininess;
#ifdef USE_PRIMITIVE
    vec2 points[4];
    uint colors[8];
    vec2 uvs[4];
#else
    vec4 modulation;
    vec4 ninepatch_margins;
    vec4 dst_rect; //for built-in rect and UV
    vec4 src_rect;
    vec2 color_texture_pixel_size;
    uint pad[6];
#endif
 
} draw_data;
```

This code fails to work (always indexes element zero no matter what, according to RenderDoc):
```
vec2 vertex = draw_data.points[gl_VertexIndex];
vec2 uv = draw_data.uvs[gl_VertexIndex];
vec4 color = vec4(unpackHalf2x16(draw_data.colors[gl_VertexIndex*2+0]),unpackHalf2x16(draw_data.colors[gl_VertexIndex*2+1]));

```

However, this does work, which does not make any sense:
```
 vec2 vertex;
    vec2 uv;
    vec4 color;
 
    if (gl_VertexIndex==0) {
        vertex = draw_data.points[0];
        uv = draw_data.uvs[0];
        color = vec4(unpackHalf2x16(draw_data.colors[0]),unpackHalf2x16(draw_data.colors[1]));
    } else if (gl_VertexIndex==1) {
        vertex = draw_data.points[1];
        uv = draw_data.uvs[1];
        color = vec4(unpackHalf2x16(draw_data.colors[2]),unpackHalf2x16(draw_data.colors[3]));
    } else if (gl_VertexIndex==2) {
        vertex = draw_data.points[2];
        uv = draw_data.uvs[2];
        color = vec4(unpackHalf2x16(draw_data.colors[4]),unpackHalf2x16(draw_data.colors[5]));
 
    } else {
        vertex = draw_data.points[3];
        uv = draw_data.uvs[3];
        color = vec4(unpackHalf2x16(draw_data.colors[6]),unpackHalf2x16(draw_data.colors[7]));
    }
```

Is there anything I'm missing? Is this a GLSLang bug or a Radeon driver bug? Any way I can find out?

**EDIT:** As mentioned below, this is illegal per-spec. I believe the compiler should generate an error and not allow it.

**Comment by darksylinc** [comment]
I'm afraid this is not a bug. I just realized 30 mins ago.

As per [the spec](https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_vulkan_glsl.txt):

> A push-constant array can only be indexed with dynamically uniform indexes.

gl_VertexIndex is not dynamically uniform, hence this code is against the spec. You can't index push constants like that.

The main advantage of push constants (vs UBOs) is that the shader starts with the values preloaded into SGPR registers, instead of being loaded from memory into SGPR during shader execution.

As such, you can't index SGPR registers with a dynamic index

**Comment by reduz** [comment]
Oh I see, that makes complete sense, so I guess this is then an actual bug in the compiler for not warning about it?

**Comment by darksylinc** [comment]
Yes you're right, it can still be seen as a glslang bug as it should catch this invalid usage as an error (not just a warning).

**Comment by johnkslang** [comment]
A dynamic index and a dynamically uniform index are two different things.  "Dynamic" is basically about a variable index rather than a constant index.  "Dynamically uniform" is about multiple invocations happening to have the same dynamic value during execution.

Front-end compilers don't know whether at runtime a variable will by dynamically uniform.

**Comment by AllianceDarkSylinc** [comment]
> Front-end compilers don't know whether at runtime a variable will by dynamically uniform.

They can track it however. Anything coming from an uniform variable has a uniform attribute (which means it's also dynamically uniform).

There's 4 types of variables:

1. Uniform
2. Threadgroup-uniform
3. SIMD/Dynamically-uniform
4. Dynamic

Each level is a downgrade from the previous one.

Variables like gl_WorkGroupID are threadgroup-uniform by nature.

Any variable coming out from anyInvocationARB and family become dynamically-uniform.

Uniform variables mixed up with dynamically-uniform variables get downgraded to dynamically-uniform
Dynamically-uniform variables mixed up with dynamic variables get downgraded to dynamic.
"Mixing" can be addition, subtraction, multiplication, etc.

Any operation where two or more variables are involved with different attributes result in a variable that has the lowest attribute.

When a variable of lower 'uniformness' is used on a function that strictly requires a variable of higher 'uniformness' variable (e.g. using a dynamic variable on a function that requires a dynamic-uniform, threadgroup-uniform or uniform argument), it should throw an error.

I think there may be other tickets in this tracker that could be closed if such feature were to be implemented. For example it is illegal to call most memory barrier variants while inside a branch that evaluates anything that isn't threadgroup-uniform or better.
FXC errors out when it detects this happens.

Metal has added the "uniform<T>" structure (where T can be int, float, etc) which is meant to be used for dynamically-uniform operations and throwing compiler errors. The interface is a bit clunky to use though.

LLVM [has a pending patch regarding a convergent](https://reviews.llvm.org/D26348) attribute for variables. I participated in the discussion to encourage its inclussion (unfortunately many developers do not have a GPU background to understand what convergent/dynamically-uniform means), but unfortunately I think my participation in that discussion only shed more darkness than light.

The context of that patch was that the GLSL optimizer was either being too aggressive (resulting in bad code) or could not perform an optimization (resulting in inefficient code).
To be more specific, Mesa drivers were optimizing this:

```
if( dynamic_variable == 0 )
{
     val = texture( myTex[0], uv );
}
else if( dynamic_variable == 1 )
{
     val = texture( myTex[1], uv );
}
```

to:

```
val = texture( myTex[dynamic_variable], uv );
```

which produced wrong GCN code (the GCN code was valid, but it used a random lane from dynamic_variable).

IIRC unfortunately this problem was resolved by making the optimizer less aggressive.

I feel this problem keeps getting delayed "to be handled later" over years by Khronos (glslang), Mesa, Apple, and now Microsoft (DXIL, see https://github.com/Microsoft/DirectXShaderCompiler/issues/1306).

It's time it starts getting attention to get it addressed, because we're all building our tech over LLVM now and IMO this attribute is vital to keep advancing the tech forward.

**Comment by AllianceDarkSylinc** [comment]
I forgot one simple case: When a user fetches data from memory that is dynamic by nature, but the user knows beforehand that this data is dynamic-uniform or better.

If that's the case, then readInvocationARB can be used to turn the dynamic variable into dynamic-uniform. If the data is not guaranteed to be dynamic uniform, the code won't work as expected but that's because we (ther user) broke the contract.

**Comment by darksylinc** [comment]
EDIT: Sorry for spamming your notifications. I realized I posted from the wrong github account, so I'm reposting from the right one.

> Front-end compilers don't know whether at runtime a variable will by dynamically uniform.

They can track it however. Anything coming from an uniform variable has a uniform attribute (which means it's also dynamically uniform).

There's 4 types of variables:

1. Uniform
2. Threadgroup-uniform
3. SIMD/Dynamically-uniform
4. Dynamic

Each level is a downgrade from the previous one.

Variables like gl_WorkGroupID are threadgroup-uniform by nature.

Any variable coming out from anyInvocationARB and family become dynamically-uniform.

Uniform variables mixed up with dynamically-uniform variables get downgraded to dynamically-uniform
Dynamically-uniform variables mixed up with dynamic variables get downgraded to dynamic.
"Mixing" can be addition, subtraction, multiplication, etc.

Any operation where two or more variables are involved with different attributes result in a variable that has the lowest attribute.

When a variable of lower 'uniformness' is used on a function that strictly requires a variable of higher 'uniformness' variable (e.g. using a dynamic variable on a function that requires a dynamic-uniform, threadgroup-uniform or uniform argument), it should throw an error.

I think there may be other tickets in this tracker that could be closed if such feature were to be implemented. For example it is illegal to call most memory barrier variants while inside a branch that evaluates anything that isn't threadgroup-uniform or better.
FXC errors out when it detects this happens.

Metal has added the "uniform<T>" structure (where T can be int, float, etc) which is meant to be used for dynamically-uniform operations and throwing compiler errors. The interface is a bit clunky to use though.

LLVM [has a pending patch regarding a convergent](https://reviews.llvm.org/D26348) attribute for variables. I participated in the discussion to encourage its inclussion (unfortunately many developers do not have a GPU background to understand what convergent/dynamically-uniform means), but unfortunately I think my participation in that discussion only shed more darkness than light.

The context of that patch was that the GLSL optimizer was either being too aggressive (resulting in bad code) or could not perform an optimization (resulting in inefficient code).
To be more specific, Mesa drivers were optimizing this:

```
if( dynamic_variable == 0 )
{
     val = texture( myTex[0], uv );
}
else if( dynamic_variable == 1 )
{
     val = texture( myTex[1], uv );
}
```

to:

```
val = texture( myTex[dynamic_variable], uv );
```

which produced wrong GCN code (the GCN code was valid, but it used a random lane from dynamic_variable, resulting in visual artifacts).

IIRC unfortunately this problem was resolved by making the optimizer less aggressive.

I feel this problem keeps getting delayed "to be handled later" over years by Khronos (glslang), Mesa, Apple, and now Microsoft (DXIL, see https://github.com/Microsoft/DirectXShaderCompiler/issues/1306).

It's time it starts getting attention to get it addressed, because we're all building our tech over LLVM now and IMO this attribute is vital to keep advancing the tech forward.

I forgot one simple case: When a user fetches data from memory that is dynamic by nature, but the user knows beforehand that this data is dynamic-uniform or better.

If that's the case, then readInvocationARB can be used to turn the dynamic variable into dynamic-uniform. If the data is not guaranteed to be dynamic uniform, the code won't work as expected but that's because we (ther user) broke the contract.

Cheers

**Comment by johnkslang** [comment]
> ...Any operation where two or more variables are involved with different attributes result in a variable that has the lowest attribute....

That sounds more like a "statically uniform" feature, but we've really chosen for "dynamically uniform".

With dynamically uniform you can "upgrade" as well:

dyn_uniform int x;
non_uniform int y;

z = x + y;  // non-uniform
w = z - y;  //  dyn_uniform

If you want to propose that we switch to a more statically-detectable feature, that probably needs to be discussed in a language spec. repo.

**Comment by darksylinc** [comment]
> If you want to propose that we switch to a more statically-detectable feature, that probably needs to be discussed in a language spec. repo.

Yes, I don't know if this is the correct place. Let me know where this discussion should be started.
And yes, I'm talking about something that is probably at language level, and would benefit optimization and detect errors statically / at compile time.

> With dynamically uniform you can "upgrade" as well:
>
>  dyn_uniform int x;
> non_uniform int y;
>
> z = x + y; // non-uniform
> w = z - y; // dyn_uniform

While it's true that these upgrades are possible, it sounds too much of a stretch, not to mention that it does not apply to floating point (a + NaN - NaN != a; likewise 0.000001 + 10000000 - 10000000 != 0.000001).

It is something that may be useful to an optimizing compiler (e.g. if the compiler can prove w is dynamically-uniform, it can perform additional optimizations), but should not be done implicitly at a language level (if w is used in a function that *requires* dynamic-uniform arguments, then w must be marked explicitly as dynamic-uniform otherwise it's a compiler error).
It is also useful to an IDE/assist tool e.g. "Did you know variable 'w' can be marked as dynamic-uniform? Because on line 123 you do x + y; then on line 124 you do (x + y) - y" like Clang analyzer currently does (e.g. it tells you a nullptr dereference is possible because "branch at line 123, entered from branch at line 64", etc).

Basically downgrading is implicit, upgrading must be explicit.
Otherwise bugs will crawl like crazy because code that relies on implicit upgrades will only work fine depending on how good the compiler is at detecting them (it behaves more like an heuristic), while downgrades are algorithmically deterministic and have well defined rules.

**Comment by johnkslang** [comment]
Upgrade can be as simple as a texture lookup where the author knows active invocations will get the same value.

Yes, we've investigated language features like you are suggesting.  For SPIR-V level, discuss at https://github.com/KhronosGroup/SPIRV-Headers for GLSL, https://github.com/KhronosGroup/GLSL.
